### PR TITLE
Add HDT Diff method

### PR DESCRIPTION
**Pull request by ate47**
This pull request ask to add the new methods, HDTDiff and HDTDiffBits and export the Bitmaps to the API.

We can describe the HDTDiff with HDTCat, If we say that the HDTCat of 2 HDT a and b is the HDT a U b, then the HDTDiff of 2 HDT can be described as the HDT a - b.

### HDTManager

In the HDTManager, we now have 2 methods to run HDTDiff:

```java
public static HDT diffHDT(String hdtFileName1, String hdtFileName2, HDTOptions hdtFormat, ProgressListener listener) throws IOException;
```

Create the diff of the HDT in hdtFileName1 - the HDT in hdtFileName2 inside the first HDT, the new HDT can be tune with the hdtFormat parameter.

```java
public static HDT diffHDTBit(String location, String hdtFileName, Bitmap deleteBitmap, HDTOptions hdtFormat, ProgressListener listener) throws IOException;
```

Create a new HDT with location location from the HDT in hdtFileName minus the triples with position (see #150) marked on the delete bitmap, the new HDT can be tune with the hdtFormat parameter.

To use this method, I've moved the Bitmap classes from the CORE to the API.

### Bitmaps

We now have 3 new class:

- ``org.rdfhdt.hdt.compact.bitmap.Bitmap`` (Previously in the CORE, moved to the API)
- ``org.rdfhdt.hdt.compact.bitmap.ModifiableBitmap`` (Previously in the CORE, moved to the API)
- ``org.rdfhdt.hdt.compact.bitmap.BitmapFactory`` (Previously in the CORE, moved and updated to the API)

The bitmaps are the same as the previous ones in the CORE, I've updated the Javadoc to have a description on each method. In the BitmapFactory, we keep the same methods, but with the addition of some method to control the created Bitmap:

```java
public static Bitmap createEmptyBitmap(long size);
```

Create an empty unmodifiable bitmap, all the accesses are the same with a bitmap filled with 0.

```java
public static ModifiableBitmap createRWBitmap(long size);
```

Create a ModifiableBitmap with only the Bitmap.access(long), Bitmap.countOnes(), Bitmap.countZeros(), ModifiableBitmap.set(long, boolean), Bitmap.getNumBits(), Bitmap.getSizeBytes() and ModifiableBitmap.append(boolean) operations, so no rank and select need to be compute, avoiding to have to build an index. Useful for the diffHDTBit that doesn't require an indexed Bitmap.

### HDT Diff of 2 HDTs

We build a delete bitmap for the first HDT to remove all the triples of HDT2 in HDT1, for example, if we have:

**HDT1**

```
ex:aaa ex:pred ex:aaa .
ex:aaa ex:pred ex:bbb .
ex:aaa ex:pred ex:ccc .
```

and 

**HDT2**

```
ex:aaa ex:pred ex:aaa .
ex:ddd ex:pred ex:ccc .
```

We are creating the delete bitmap ``100``, because only the 1st triple should be removed from HDT1 ``ex:aaa ex:pred ex:aaa``.

Then the HDTDiff is calling the HDTDiffBit with HDT1 and this new delete bitmap.

### HDT Diff of a HDT and a delete bitmap

We create a new HDT from the HDT with all the triple marked to be deleted deleted. HDTDiffBit will also recompute the shared subjects/objects to remove from shared elements, elements that aren't shared anymore, the Dictionary is preserved.

For example, if we have the HDT

```
Dict:
  Shared: ex:aaa ex:ccc
  Subj: ex:eee
  Pred: ex:pred
  Obj:  ex:bbb ex:ddd ex:fff
Triples:
  ex:aaa ex:pred ex:aaa .
  ex:aaa ex:pred ex:bbb .
  ex:aaa ex:pred ex:ccc .
  ex:ccc ex:pred ex:ccc .
  ex:ccc ex:pred ex:ddd .
  ex:eee ex:pred ex:fff .
```

and the delete bitmap ``0101010``, we will delete the triples:

```
ex:aaa ex:pred ex:bbb .
ex:ccc ex:pred ex:ccc .
```

The result HDT would be:

```
Dict:
  Shared: ex:aaa
  Subj: ex:eee
  Pred: ex:pred
  Obj:  ex:ccc ex:ddd ex:fff
Triples:
  ex:aaa ex:pred ex:aaa .
  ex:aaa ex:pred ex:ccc .
  ex:ccc ex:pred ex:ddd .
  ex:eee ex:pred ex:fff .
```

Notice that ex:ccc isn't shared anymore.

## Tests

To test this features, I've added 4 tests files:

- ``org.rdfhdt.hdt.compact.bitmap.BitmapFactoryTest``, test be can use the BitmapFactory and the new EmptyBitmap class.
- ``org.rdfhdt.hdt.triples.impl.BitmapTriplesIteratorDiffTest``, test the Iterator to build the delete bitmap from 2 HDTs.
- ``org.rdfhdt.hdt.hdtDiff.HdtDiffTest``, test the HDTDiff of generated HDT.
- ``org.rdfhdt.hdt.hdtDiff.HdtDiffStaticTest``, test the HDTDiff of nt files.

For each test file, I'm testing the HDT method for the 4 dictionary types:

- ``<http://purl.org/HDT/hdt#dictionaryFour>``
- ``dictionaryMultiObj``
- ``dictionaryFourBig``
- ``<http://purl.org/HDT/hdt#dictionaryFourPsfc>``

For HDTDiff test, we are also checking if the asked dictionary is the same as the result HDTs.

### CODE REVIEW

File: hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap.java [hunk 1]
```diff
@@ -40,84 +40,95 @@
 public interface Bitmap {
 	/**
 	 * Get the value of the bit at position pos
-	 * @param pos
+	 * @param pos pos
```
Comment (D063520) [inline hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap.java:43]: should this not be only one "pos"

File: hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/BitmapFactory.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+public abstract class BitmapFactory {
+
+    private static BitmapFactory instance;
+
+    private static BitmapFactory getInstance() {
+        if (instance == null) {
+            try {
+                // Try to instantiate pro
+                Class<?> managerImplClass = Class.forName("org.rdfhdt.hdt.pro.BitmapFactory");
+                instance = (BitmapFactory) managerImplClass.newInstance();
```
Comment (D063520) [inline hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/BitmapFactory.java:21]: what is the pro name space? is it new?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap64.java [hunk 3]
```diff
@@ -50,7 +51,7 @@
  * @author mario.arias
  *
  */
-public class Bitmap64 {
+public class Bitmap64 implements ModifiableBitmap{
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap64.java:54]: can you explain why you had to do this? It's not so clear to me ....

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/EmptyBitmap.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import org.rdfhdt.hdt.util.crc.CRC32;
+import org.rdfhdt.hdt.util.crc.CRC8;
+import org.rdfhdt.hdt.util.crc.CRCOutputStream;
+import org.rdfhdt.hdt.util.io.IOUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class EmptyBitmap implements ModifiableBitmap {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/EmptyBitmap.java:16]: what is the reason to create this class?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import org.rdfhdt.hdt.compact.bitmap.ModifiableBitmap;
+import org.rdfhdt.hdt.dictionary.impl.utilCat.CatMapping;
+import org.rdfhdt.hdt.listener.ProgressListener;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public interface DictionaryDiff {
+    void diff(Dictionary dictionary, Map<String, ModifiableBitmap> bitmaps, ProgressListener listener) throws IOException;
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java:12]: why is bitmaps a map?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryFactory.java [hunk 6]
```diff
@@ -103,4 +104,14 @@ public static DictionaryPrivate createDictionary(ControlInfo ci) {
 		}
 		throw new IllegalFormatException("Implementation of dictionary not found for "+name);
 	}
+
+	public static DictionaryDiff createDictionaryDiff(Dictionary dictionary, String location) {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryFactory.java:108]: coudl you document this method? For me it is not clear what location is ... the location which is used to compute the diff or the location of a second hdt?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/FourSectionDictionaryDiff.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                if (type == 1) {
+                    mappingBack.set(allMappings.get("subject").getMapping(i) - 1, (i + 1 + (int) dictionary.getNshared()), 1);
+                } else if (type == 2) {
+                    mappingBack.set(allMappings.get("subject").getMapping(i) + numShared - 1, (i + 1 + (int) dictionary.getNshared()), 2);
+                }
+            }
+        }
+    }
+
+    private static class SharedWrapper implements Iterator<CatElement> {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/FourSectionDictionaryDiff.java:181]: I'm not sure ... but at a first sight it seams that this code is duplicated in [MultipleSectionDictionaryDiff.java](https://github.com/rdfhdt/hdt-java/pull/153/files#diff-f31a5504b73e5296134f5692c7985dfd82481397513d484db95da1245b24e841)

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
+package org.rdfhdt.hdt.dictionary.impl.utilDiff;
+
+import org.rdfhdt.hdt.compact.bitmap.Bitmap;
+import org.rdfhdt.hdt.dictionary.impl.utilCat.CatElement;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+
+public class DiffWrapper implements Iterator<CatElement> {
+
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java:10]: can you put some documentation here to explain this class ?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package org.rdfhdt.hdt.triples.impl;
+
+import org.rdfhdt.hdt.compact.bitmap.ModifiableBitmap;
+import org.rdfhdt.hdt.exceptions.NotFoundException;
+import org.rdfhdt.hdt.hdt.HDT;
+import org.rdfhdt.hdt.triples.IteratorTripleString;
+import org.rdfhdt.hdt.triples.TripleString;
+
+public class BitmapTriplesIteratorDiff {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java:9]: Here I would add "Class that computes the triples that returns a bitmap marking the triples in hdtOriginal that are also in hdtDiff"

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import org.rdfhdt.hdt.enums.ResultEstimationType;
+import org.rdfhdt.hdt.hdt.HDT;
+import org.rdfhdt.hdt.triples.BitmapTriplesIteratorDiffBit;
+import org.rdfhdt.hdt.triples.IteratorTripleID;
+import org.rdfhdt.hdt.triples.TripleID;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class BitmapTriplesIteratorDiffBitImpl implements BitmapTriplesIteratorDiffBit {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java:16]: some documentation also here ....

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java (hunk 5 above)
Comment (D063520) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java:12]: ahhh, there is one for each section .... I would document this ...

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/BitmapTriplesIteratorDiffBit.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import org.rdfhdt.hdt.dictionary.Dictionary;
+import org.rdfhdt.hdt.exceptions.IllegalFormatException;
+import org.rdfhdt.hdt.hdt.HDT;
+import org.rdfhdt.hdt.hdt.HDTVocabulary;
+import org.rdfhdt.hdt.triples.impl.BitmapTriplesIteratorDiffBitImpl;
+import org.rdfhdt.hdt.triples.impl.BitmapTriplesIteratorDiffBitLiterals;
+
+import java.util.Map;
+
+public interface BitmapTriplesIteratorDiffBit {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/BitmapTriplesIteratorDiffBit.java:14]: some documentation .... it should be clear what the class is doing

Comment (D063520) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/BitmapTriplesIteratorDiffBit.java:14]: Class that computes for each dictionary section the entries that can be deleted. I would maybe call it "DictionaryEntriesDiff"
Description: Computes for each dictionary section a bitmap indicating if the corresponding entry needs to be deleted or not....

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        this.hdtDiff = hdtDiff;
+        this.bitmap = bitmap;
+    }
+
+    /**
+     * fill the diff bitmap
+     */
+    public void fillBitmap() {
+        IteratorTripleString iterator1;
+        try {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java:25]: // iterate over hdtOriginal

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            throw new RuntimeException(e);
+        }
+        while (iterator1.hasNext()) {
+            TripleString origNext = iterator1.next();
+            try {
+                IteratorTripleString otherTriple = hdtDiff.search(
+                        origNext.getSubject(),
+                        origNext.getPredicate(),
+                        origNext.getObject()
+                );
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiff.java:37]: // if the triple is found also in the hdtDiff then mark it as to delete

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        this.bitArrayDisk = deleteBitmap;
+        this.bitmaps = new HashMap<>();
+    }
+
+    @Override
+    public void loadBitmaps() {
+        this.bitmaps.clear();
+        count = 0;
+
+        Dictionary dict = hdtOriginal.getDictionary();
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java:36]: // create a bitmap for each section

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        this.bitmaps.put("P", BitmapFactory.createRWBitmap(dict.getPredicates().getNumberOfElements()));
+        this.bitmaps.put("S", BitmapFactory.createRWBitmap(dict.getSubjects().getNumberOfElements()));
+        this.bitmaps.put("O", BitmapFactory.createRWBitmap(dict.getObjects().getNumberOfElements()));
+        this.bitmaps.put("SH_S", BitmapFactory.createRWBitmap(dict.getShared().getNumberOfElements()));
+        this.bitmaps.put("SH_O", BitmapFactory.createRWBitmap(dict.getShared().getNumberOfElements()));
+
+        while (iterator.hasNext()) {
+            TripleID tripleID = iterator.next();
+            count = iterator.getLastTriplePosition();
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java:45]: // if triple is marked as to delete

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java (hunk 10 above)
Comment (D063520) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitImpl.java:16]: I would name this "FourSectionDictionaryEntriesDiff" or something similar ...

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitLiterals.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+import org.rdfhdt.hdt.hdt.HDT;
+import org.rdfhdt.hdt.triples.BitmapTriplesIteratorDiffBit;
+import org.rdfhdt.hdt.triples.IteratorTripleID;
+import org.rdfhdt.hdt.triples.TripleID;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class BitmapTriplesIteratorDiffBitLiterals implements BitmapTriplesIteratorDiffBit {
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorDiffBitLiterals.java:17]: I would name this "MultiSectionDictionaryEntriesDiff" or something similar ...

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/FourSectionDictionaryDiff.java (hunk 7 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/FourSectionDictionaryDiff.java:181]: Many chances are made in the diff method

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/EmptyBitmap.java (hunk 4 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/EmptyBitmap.java]: It was to allow the usage of empty bitmap, but it was only in tests, I've removed it

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap64.java (hunk 3 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/compact/bitmap/Bitmap64.java:54]: To be able to use all the Bitmap64 methods without having to use an index like with Bitmap375

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryFactory.java (hunk 6 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryFactory.java:113]: I've updated the javadoc in [this commit](https://github.com/ate47/hdt-java/commit/9aae06eb43b244cd94a41888999651b990e673a1) to remove this issue

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java (hunk 5 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/DictionaryDiff.java:19]: I've added some javadoc for this class, is it ok?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java (hunk 8 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java:13]: I've added some javadoc for this class, is it ok?

File: hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/BitmapFactory.java (hunk 2 above)
Comment (ate47) [reply hdt-api/src/main/java/org/rdfhdt/hdt/compact/bitmap/BitmapFactory.java:21]: To get the same possibilities of the [HDTManager](https://github.com/ate47/hdt-java/blob/diff_func/hdt-api/src/main/java/org/rdfhdt/hdt/hdt/HDTManager.java#L24) ``getInstance()`` method

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java (hunk 8 above)
Comment (D063520) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/dictionary/impl/utilDiff/DiffWrapper.java:13]: market - > marked

**Pull request merged**